Make j.u.HashMap loadFactor constructor follow Java 8 description.

LeeTibbert opened the pull request:
This PR changes j.u.HashMap to follow the the Java 8 documentation for [HashMap](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html). The loadFactor constructor argument as is described as float.

The type of loadFactor in the private class NullRejectingHashMap, which extends HashMap,
was changed to match.

sjrd reviewed (changes requested):
Thank you for the PR! Some comments:

* We should add an `@Test` in `HashMapTest` to make sure that we can now use that constructor (no need to do a full new factory)
* For clarity, it would be good to add an explicit `.toDouble` when we use `loadFactor` in `newThreshold`
* Could you also rebase on top of the latest master (we just had a release), and remove the last paragraph from the commit message (as it makes no sense in the git history, although it's nice to know on the human side).

LeeTibbert commented:
Thank you for the rapid turn around, especially just before & immediately after a release:

1) Commit message amended as suggested

2) re: load factor math.
```
@inline private[this] def newThreshold(size: Int): Int =
    (size.toDouble * loadFactor).toInt
```
  I show my lack of understanding of compiler internals.   I can do the suggested `loadFactor.toDouble`.
 Given that size is known to be an Int it seems that ` (size * loadFactor).toInt` or even
 `(size.toFloat * loadFactor).toInt would be more idiomatic.   

 The former is what I would naturally write, perhaps in error.  The worst that happens with
  single precision math is an off by one error; unlikely since the intermediate result will be truncated.

  Is there some difference in the number of instructions or the cost of the instructions at issue time
  in the promotion of the arguments?   Is it a matter of a human reader of the code?

3) re: HashMapTest  I can change the name as directed or provide a better test.
   
    As long as HashMapFactory is fufilling its contract, calling factory.empty should have caused factory
    to invoke one of the constructors of class HashMap, which should have exercised both the initialCapacity
    and loadFactor arguments.   This seemed to be the least intrusive way to slip the new test into existing code.
    
     I believe that *Test classes which extend HashMap will also exercise the loadFactor argument, but 
     the new test means that the argument will get exercised whenever HashMapTest is executed; self-contained
     with no need to rely upon the existence of extending tests.
 
      Of course, there is always the argument that the new test is redundant.



Thank you, I am trying to increase my understanding & proficiency so that my code takes less time
to review.

sjrd commented:
For 2), the issue is that not all `Int`s can be accurately converted to a `Float`. So an implicit or explicit `size.toFloat` can already introduce a significant error. `Double`s do not have that problem: all `Int`s can be accurately converted to `Double`.

For 3), the issue is that, although you're right that the constructor is transitively called through the other constructors, that does not exercise whether the signature is the correct binary signature to be called from user code. To test that, we need to call that constructor directly from the test suite, not indirectly. A correct test for that would fail to link prior to fixing the signature from `Double` to `Float`.

LeeTibbert commented:
Thank you for taking the teachable moment.

> all Ints can be accurately converted to Double
I'll add that to my memory and numerical issues to guard against.

LeeTibbert commented:
I added a test to HashMapSuite.scala

1) Since it is only one line, I removed the braces.  Should I have kept them?

2) When I examine the log file, the test gets executed 5 times, because of
     other tests inheriting from it.  Each of the extending classes repeats the
     test against j.u.HashMap instead of its own type.  Those later tests
     appear to be a waste. Seems like I need to  change something to either restrict 
     the test to HashMap  or modifying the extending Test classes.

     Is adding an override in each of the extending classes which calls into
     a new Factory for that class the way to go? Is there something simpler?

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/HashMapTest.scala:
```diff
@@ -16,8 +16,14 @@ import java.{util => ju}
   def factory: HashMapFactory = new HashMapFactory
+
+  @Test def testConstructorWithLoadFactorArg(): Unit =
+    assertNotNull(new ju.HashMap[Int,String](10, 0.5f))
```
```suggestion
    assertNotNull(new ju.HashMap[Int, String](10, 0.5f))
```
The `assertNotNull` doesn't really do anything. If the result of `new` is `null`, we have bigger problems than a bad library definition. You might as well remove it.

A comment mentioning that is this mostly to test that the constructor links would be very useful here, though.

sjrd reviewed (changes requested):
> Since it is only one line, I removed the braces. Should I have kept them?

For any method than an `@Test`, removing them is OK. For an `@Test` method, it looks a bit strange. Ideally keep the braces here, although it's not strictly necessary.

> When I examine the log file, the test gets executed 5 times, because of
other tests inheriting from it. Each of the extending classes repeats the
test against j.u.HashMap instead of its own type.

That seems unavoidable given our architecture. We'll just have to live with it.

LeeTibbert commented:
Build failure appears unrelated to the content of this PR.  Other pending PRs have the same build error.

sjrd commented:
You'll need to rebase on the latest master to solve the CI failure. Could you also squash everything into a single commit with a clean commit message?

LeeTibbert commented:
Re-based to Repository as of late 2020-10-20, edited commit comment, & squashed.  
CI is all green except for two failures (one each Scala 2.11 & Scala 2.12) in tools-sbtplugin.
`[error] java.lang.RuntimeException: Scala.js linker interface: Binary compatibility check failed!`
Reading the log files, it looks like the error is unrelated to the changes of this PR.

I _think_ I got the rebase correct.  Should I try again?  Start over with a new PR, sure to have
most current Repository?  Please advise. Thank you.
